### PR TITLE
SYNR-1171

### PR DESCRIPTION
**Pull request by kimyen**

### CODE REVIEW

File: src/CastPyObjects.c [hunk 1]
```diff
@@ -101,100 +207,94 @@ SEXP py_class(PyObject *py_object){
...
+    PyCollection_AllSameType
+      params
+        py_object
+          the python collection to check
+        collection_type
+          the type of py_object,
+          use LIST_COLLECTION_TYPE or TUPLE_COLLECTION_TYPE
+      returns
+        the integer that Py_GetR_Type() returns for all items in the list (except None),
+        or -1 if items in the list do not have the same type.
```
Comment (kimyen) [inline src/CastPyObjects.c]: or R_UNCATEGORIZED_TYPE

File: tests/testthat/test_pyGet.R [hunk 2]
```diff
@@ -2,54 +2,473 @@
...
+})
+
+test_that("dict with None key can be converts to named vector with NA name in r", {
+  pyExec("py_dict = {None: True, 'b':False}")
+  expect_output(pyExecp("type(py_dict)"), "dict")
+  r_value <- pyGet("py_dict")
+  expect_equal("logical", class(r_value))
+  expected <- c(TRUE, FALSE)
+  names(expected) <- c(NA, 'b')
+  expect_equal(expected['b'], r_value['b'])
```
Comment (brucehoff) [inline tests/testthat/test_pyGet.R]: expected[is.na(names(expected))] should equal TRUE

File: src/CastPyObjects.c [hunk 3]
```diff
@@ -10,6 +10,112 @@
 #include "PythonInR.h"
 #include "CToR.h"
 
+const int LIST_COLLECTION_TYPE = 0;
+const int TUPLE_COLLECTION_TYPE = 1;
+const int R_NA_TYPE = 0;
+const int R_LOGICAL_TYPE = 10;
+const int R_INTEGER_TYPE = 13;
+const int R_NUMERIC_TYPE = 14;
+const int R_CHARACTER_TYPE = 16;
```
Comment (brucehoff) [inline src/CastPyObjects.c]: avoid redefining values already defined in Rinternals.h

File: src/CastPyObjects.c [hunk 4]
```diff
@@ -101,100 +207,95 @@ SEXP py_class(PyObject *py_object){
...
-int PyTuple_AllSameType(PyObject *py_object){
-    PyObject *item, *py_len, *py_i;
-    long list_len, count;
-    int r_type = -1, item_type;
+    if (collection_type == LIST_COLLECTION_TYPE) {
+      py_len = PyLong_FromSsize_t(PyList_GET_SIZE(py_object));
+    } else if (collection_type == TUPLE_COLLECTION_TYPE) {
+      py_len = PyLong_FromSsize_t(PyTuple_GET_SIZE(py_object));
+    } else {
+      // should throw an error here
```
Comment (brucehoff) [inline src/CastPyObjects.c]: throw error here (call the error throwing function)

**Pull request merged**